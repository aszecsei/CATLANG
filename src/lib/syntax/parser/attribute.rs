use toolshed::list::GrowableList;

use crate::syntax::ast::*;
use crate::syntax::error::*;
use crate::syntax::lexer::Token;
use crate::syntax::parser::Parser;

impl<'ast> Parser<'ast> {
    pub fn attribute_node(&mut self) -> Result<AttributeNode<'ast>> {
        unimplemented!()
    }
}
