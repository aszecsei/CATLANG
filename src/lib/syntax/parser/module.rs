use super::*;

impl<'ast> Parser<'ast> {
    pub fn import(&mut self) -> Result<ImportNode<'ast>> {
        let start = self.start_then_advance();
        let import_list: ImportList = self.import_list()?;

        self.expect(Token::From);
        let path = self.string_literal_node()?;
        self.eat(Token::Semicolon);

        Ok(self.node_at(start, path.end, Import { import_list, path }))
    }

    fn import_list(&mut self) -> Result<ImportList<'ast>> {
        if self.eat(Token::Mul) {
            // Glob import
            self.expect(Token::As);
            let identifier = self.identifier_node()?;
            Ok(GlobImportList { identifier }.into())
        } else {
            // Named import
            self.expect(Token::LCurlyB);
            let names = GrowableList::new();
            loop {
                match self.current_token {
                    Token::RCurlyB => break,
                    _ => {
                        let identifier = self.identifier_node()?;
                        let renamed_to = if self.eat(Token::As) {
                            Some(self.identifier_node()?)
                        } else {
                            None
                        };
                        let start = identifier.start;
                        let end = if let Some(renamed) = renamed_to {
                            renamed.end
                        } else {
                            identifier.end
                        };
                        names.push(
                            self.arena,
                            self.node_at(
                                start,
                                end,
                                ImportIdentifier {
                                    identifier,
                                    renamed_to,
                                },
                            ),
                        );
                        self.expect_one_of(&[Token::Comma], &[Token::RCurlyB]);
                    }
                }
            }
            self.expect(Token::RCurlyB);
            Ok(NamedImportList {
                imports: names.as_list(),
            }
            .into())
        }
    }

    pub fn export(&mut self) -> Result<ExportNode<'ast>> {
        let start = self.start_then_advance();
        if is_declaration_starter(self.current_token) {
            // Export declaration
            let declaration = self.declaration_node()?;
            Ok(self.node_at(start, declaration.end, declaration))
        } else if self.current_token == Token::LCurlyB || self.current_token == Token::Mul {
            // Re-export
            let exports = self.import_list()?;
            self.expect(Token::From);
            let path = self.string_literal_node()?;
            self.eat(Token::Semicolon);
            Ok(self.node_at(start, path.end, ExportReExport { exports, path }))
        } else {
            // Export statement
            let identifier = self.identifier_node()?;
            let renamed_to = if self.eat(Token::As) {
                Some(self.identifier_node()?)
            } else {
                None
            };
            let end = if let Some(renamed) = renamed_to {
                renamed.end
            } else {
                identifier.end
            };
            self.eat(Token::Semicolon);
            Ok(self.node_at(
                start,
                end,
                ExportStatement {
                    identifier,
                    renamed_to,
                },
            ))
        }
    }
}

pub fn is_declaration_starter(t: Token) -> bool {
    matches!(
        t,
        Token::Const | Token::Type | Token::Let | Token::Function | Token::Struct | Token::Enum
    )
}

#[cfg(test)]
mod tests {
    use super::*;
    use insta::assert_debug_snapshot;
    use toolshed::Arena;

    #[test]
    fn test_glob_import() {
        let source = "import * as vectors from \"./vectors\"";
        let arena = Arena::new();
        let mut p = Parser::new(source, &arena);
        let res = p.import().unwrap();

        assert_debug_snapshot!(res);
    }

    #[test]
    fn test_named_imports() {
        let source = "import { Vector2 as Vec2, Vector3 } from \"vectors\";";
        let arena = Arena::new();
        let mut p = Parser::new(source, &arena);
        let res = p.import().unwrap();

        assert_debug_snapshot!(res);
    }

    #[test]
    fn test_glob_re_export() {
        let source = "export * as vectors from \"vectors\";";
        let arena = Arena::new();
        let mut p = Parser::new(source, &arena);
        let res = p.export().unwrap();

        assert_debug_snapshot!(res);
    }

    #[test]
    fn test_named_re_export() {
        let source = "export { Vector2 as Vec2, Vector3 } from \"vectors\";";
        let arena = Arena::new();
        let mut p = Parser::new(source, &arena);
        let res = p.export().unwrap();

        assert_debug_snapshot!(res);
    }

    #[test]
    fn test_export_statement() {
        let source = "export PI;";
        let arena = Arena::new();
        let mut p = Parser::new(source, &arena);
        let res = p.export().unwrap();

        assert_debug_snapshot!(res);
    }

    #[test]
    fn test_renamed_export_statement() {
        let source = "export PI_CONST as PI;";
        let arena = Arena::new();
        let mut p = Parser::new(source, &arena);
        let res = p.export().unwrap();

        assert_debug_snapshot!(res);
    }
}
