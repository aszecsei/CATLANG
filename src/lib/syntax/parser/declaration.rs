use toolshed::list::GrowableList;

use crate::syntax::ast::*;
use crate::syntax::error::*;
use crate::syntax::lexer::Token;
use crate::syntax::parser::Parser;

impl<'ast> Parser<'ast> {
    pub fn declaration_node(&mut self) -> Result<DeclarationNode<'ast>> {
        // TODO: Possible attributes before this
        let start = self.loc().0;
        let declarator = self.declarator()?;
        let end = self.last_span.end as u32;
        Ok(self.node_at(
            start,
            end,
            Declaration {
                attributes: GrowableList::new().as_list(),
                declarator,
            },
        ))
    }

    fn declarator(&mut self) -> Result<Declarator<'ast>> {
        match self.current_token {
            Token::Const => self.const_declarator(),
            Token::Type => self.type_declarator(),
            Token::Let => self.variable_declarator(),
            Token::Function => self.function_declarator(),
            Token::Struct => self.struct_declarator(),
            Token::Enum => self.enum_declarator(),
            _ => Err(Error::ExpectedOneOfButGot {
                expected_tokens: vec![
                    Token::Const,
                    Token::Type,
                    Token::Let,
                    Token::Function,
                    Token::Struct,
                    Token::Enum,
                ],
                token: self.current_token,
                raw: self.current_slice.into(),
                span: self.current_span.clone(),
            }),
        }
    }

    fn const_declarator(&mut self) -> Result<Declarator<'ast>> {
        let _start = self.start_then_advance();
        let identifier = self.identifier_node()?;
        let type_expression = if self.eat(Token::Colon) {
            Some(self.type_node()?)
        } else {
            None
        };
        self.expect(Token::Assign);
        let expression = self.expression_node()?;
        Ok(ConstantDeclarator {
            identifier,
            type_expression,
            expression,
        }
        .into())
    }

    fn type_declarator(&mut self) -> Result<Declarator<'ast>> {
        let _start = self.start_then_advance();
        let identifier = self.identifier_node()?;
        self.expect(Token::Assign);
        let type_expression = self.type_node()?;
        Ok(TypeDeclarator {
            identifier,
            type_expression,
        }
        .into())
    }

    fn variable_declarator(&mut self) -> Result<Declarator<'ast>> {
        let _start = self.start_then_advance();
        let identifier = self.identifier_node()?;
        let type_expression = if self.eat(Token::Colon) {
            Some(self.type_node()?)
        } else {
            None
        };
        let expression = if self.eat(Token::Assign) {
            Some(self.expression_node()?)
        } else {
            None
        };
        Ok(VariableDeclarator {
            identifier,
            type_expression,
            expression,
        }
        .into())
    }

    fn function_declarator(&mut self) -> Result<Declarator<'ast>> {
        let _start = self.start_then_advance();
        let function_name = self.identifier_node()?;
        let generic_parameters = GrowableList::new();
        if self.eat(Token::LessThan) {
            while self.current_token != Token::GreaterThan && self.current_token != Token::EndOfFile
            {
                generic_parameters.push(self.arena, self.identifier_node()?);
                self.expect_one_of(&[Token::Comma], &[Token::GreaterThan]);
            }
            self.expect(Token::GreaterThan);
        }
        self.expect(Token::LParen);
        let parameters = self.formal_parameter_list()?;
        self.expect(Token::RParen);
        let return_type = if self.eat(Token::Arrow) {
            match self.current_token {
                Token::LCurlyB => None,
                _ => Some(self.type_node()?),
            }
        } else {
            None
        };
        let block = self.block_node()?;

        Ok(FunctionDeclarator {
            function_name,
            generic_parameters: generic_parameters.as_list(),
            parameters,
            return_type,
            block,
        }
        .into())
    }

    fn formal_parameter_list(&mut self) -> Result<NodeList<'ast, Parameter<'ast>>> {
        let param_list = GrowableList::new();
        while self.current_token != Token::RParen && self.current_token != Token::EndOfFile {
            let identifier = self.identifier_node()?;
            self.expect(Token::Colon);
            let type_expression = self.type_node()?;
            param_list.push(
                self.arena,
                self.node_at(
                    identifier.start,
                    type_expression.end,
                    Parameter {
                        identifier,
                        type_expression,
                    },
                ),
            );
            self.expect_one_of(&[Token::Comma], &[Token::RParen]);
        }
        Ok(param_list.as_list())
    }

    fn struct_declarator(&mut self) -> Result<Declarator<'ast>> {
        let _start = self.start_then_advance();
        let identifier = self.identifier_node()?;
        let generic_parameters = GrowableList::new();
        if self.eat(Token::LessThan) {
            while self.current_token != Token::GreaterThan {
                generic_parameters.push(self.arena, self.identifier_node()?);
                self.expect_one_of(&[Token::Comma], &[Token::GreaterThan]);
            }
        }
        self.expect(Token::LCurlyB);
        let members = self.struct_member_list()?;
        self.expect(Token::RCurlyB);
        Ok(StructDeclarator {
            identifier,
            generic_parameters: generic_parameters.as_list(),
            members,
        }
        .into())
    }

    fn struct_member_list(&mut self) -> Result<NodeList<'ast, StructMember<'ast>>> {
        let member_list = GrowableList::new();
        while self.current_token != Token::RCurlyB && self.current_token != Token::EndOfFile {
            let start = self.current_span.start as u32;
            let is_owned = self.eat(Token::Owned);
            let identifier = self.identifier_node()?;
            self.expect(Token::Colon);
            let type_expression = self.type_node()?;
            let mut end = type_expression.end;
            let default_value = if self.eat(Token::Assign) {
                let expr = self.expression_node()?;
                end = expr.end;
                Some(expr)
            } else {
                None
            };
            member_list.push(
                self.arena,
                self.node_at(
                    start,
                    end,
                    StructMember {
                        identifier,
                        is_owned,
                        type_expression,
                        default_value,
                    },
                ),
            );
            self.expect_one_of(&[Token::Semicolon], &[Token::RCurlyB]);
        }

        Ok(member_list.as_list())
    }

    fn enum_declarator(&mut self) -> Result<Declarator<'ast>> {
        self.expect(Token::Enum);
        let identifier = self.identifier_node()?;
        let representation = if self.eat(Token::Colon) {
            Some(self.enum_representation()?)
        } else {
            None
        };
        self.expect(Token::LCurlyB);
        let values = self.enum_member_list()?;
        self.expect(Token::RCurlyB);

        Ok(EnumDeclarator {
            identifier,
            representation,
            values,
        }
        .into())
    }

    fn enum_representation(&mut self) -> Result<EnumRepresentationType> {
        let tok = self.current_token;
        self.bump();
        match tok {
            Token::S8 => Ok(EnumRepresentationType::S8),
            Token::U8 => Ok(EnumRepresentationType::U8),
            Token::S16 => Ok(EnumRepresentationType::S16),
            Token::U16 => Ok(EnumRepresentationType::U16),
            Token::S32 => Ok(EnumRepresentationType::S32),
            Token::U32 => Ok(EnumRepresentationType::U32),
            Token::S64 => Ok(EnumRepresentationType::S64),
            Token::U64 => Ok(EnumRepresentationType::U64),
            _ => Err(Error::ExpectedOneOfButGot {
                expected_tokens: vec![
                    Token::S8,
                    Token::U8,
                    Token::S16,
                    Token::U16,
                    Token::S32,
                    Token::U32,
                    Token::S64,
                    Token::U64,
                ],
                token: self.current_token,
                raw: self.lexer.slice().into(),
                span: self.lexer.span(),
            }),
        }
    }

    fn enum_member_list(&mut self) -> Result<NodeList<'ast, EnumValue<'ast>>> {
        let member_list = GrowableList::new();
        while self.current_token != Token::RCurlyB && self.current_token != Token::EndOfFile {
            let identifier = self.identifier_node()?;
            let mut end = identifier.end;
            let value = if self.eat(Token::Assign) {
                let expr = self.expression_node()?;
                end = expr.end;
                Some(expr)
            } else {
                None
            };
            member_list.push(
                self.arena,
                self.node_at(identifier.start, end, EnumValue { identifier, value }),
            );
            self.expect_one_of(&[Token::Comma], &[Token::RCurlyB]);
        }

        Ok(member_list.as_list())
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use insta::assert_debug_snapshot;
    use toolshed::Arena;

    #[test]
    fn test_const_declarator() {
        let source = "const PI = 3;";
        let arena = Arena::new();
        let mut p = Parser::new(source, &arena);
        let res = p.declaration_node().unwrap();

        assert_debug_snapshot!(res);
    }

    #[test]
    fn test_uninitialized_var_declarator() {
        let source = "let pi;";
        let arena = Arena::new();
        let mut p = Parser::new(source, &arena);
        let res = p.declaration_node().unwrap();

        assert_debug_snapshot!(res);
    }

    #[test]
    fn test_initialized_var_declarator() {
        let source = "let pi = 3;";
        let arena = Arena::new();
        let mut p = Parser::new(source, &arena);
        let res = p.declaration_node().unwrap();

        assert_debug_snapshot!(res);
    }

    #[test]
    fn test_type_declarator() {
        let source = "type f_array = []float";
        let arena = Arena::new();
        let mut p = Parser::new(source, &arena);
        let res = p.declaration_node().unwrap();

        assert_debug_snapshot!(res);
    }

    #[test]
    fn test_function_declarator() {
        let source = "function foo() {}";
        let arena = Arena::new();
        let mut p = Parser::new(source, &arena);
        let res = p.declaration_node().unwrap();

        assert_debug_snapshot!(res);
    }

    #[test]
    fn test_function_params_declarator() {
        let source = "function foo(i: int) {}";
        let arena = Arena::new();
        let mut p = Parser::new(source, &arena);
        let res = p.declaration_node().unwrap();

        assert_debug_snapshot!(res);
    }

    #[test]
    fn test_function_params_return_declarator() {
        let source = "function foo(i: int) -> int {}";
        let arena = Arena::new();
        let mut p = Parser::new(source, &arena);
        let res = p.declaration_node().unwrap();

        assert_debug_snapshot!(res);
    }

    #[test]
    fn test_function_params_return_generic_declarator() {
        let source = "function foo<T>(i: T) -> int {}";
        let arena = Arena::new();
        let mut p = Parser::new(source, &arena);
        let res = p.declaration_node().unwrap();

        assert_debug_snapshot!(res);
    }

    #[test]
    fn test_struct_declarator() {
        let source = "struct Vector2 { x: f32 = 0; y: f32 = 0; }";
        let arena = Arena::new();
        let mut p = Parser::new(source, &arena);
        let res = p.declaration_node().unwrap();

        assert_debug_snapshot!(res);
    }

    #[test]
    fn test_generic_struct_declarator() {
        let source = "struct Vector2<T> { x: T; y: T; }";
        let arena = Arena::new();
        let mut p = Parser::new(source, &arena);
        let res = p.declaration_node().unwrap();

        assert_debug_snapshot!(res);
    }

    #[test]
    fn test_enum_declarator() {
        let source = "enum MyEnum { A, B }";
        let arena = Arena::new();
        let mut p = Parser::new(source, &arena);
        let res = p.declaration_node().unwrap();

        assert_debug_snapshot!(res);
    }

    #[test]
    fn test_enum_values_declarator() {
        let source = "enum MyEnum { A, B = 3 }";
        let arena = Arena::new();
        let mut p = Parser::new(source, &arena);
        let res = p.declaration_node().unwrap();

        assert_debug_snapshot!(res);
    }

    #[test]
    fn test_enum_repr_declarator() {
        let source = "enum MyEnum : u8 { A, B = 3 }";
        let arena = Arena::new();
        let mut p = Parser::new(source, &arena);
        let res = p.declaration_node().unwrap();

        assert_debug_snapshot!(res);
    }
}
