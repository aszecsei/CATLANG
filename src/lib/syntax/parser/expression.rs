use crate::syntax::ast::*;
use crate::syntax::error::*;
use crate::syntax::lexer::Token;
use crate::syntax::parser::Parser;

impl<'ast> Parser<'ast> {
    pub fn expression_node(&mut self) -> Result<ExpressionNode<'ast>> {
        self.assignment_expression()
    }

    // Right-associative
    fn assignment_expression(&mut self) -> Result<ExpressionNode<'ast>> {
        let lhs = self.conditional_expression()?;
        let assignment = match self.current_token {
            Token::Assign => Some(AssignmentOperator::Plain),
            Token::AddAssign => Some(AssignmentOperator::Addition),
            Token::SubAssign => Some(AssignmentOperator::Subtraction),
            Token::MulAssign => Some(AssignmentOperator::Multiplication),
            Token::QuoAssign => Some(AssignmentOperator::Division),
            Token::ModAssign => Some(AssignmentOperator::Remainder),
            Token::BitOrAssign => Some(AssignmentOperator::BitOr),
            Token::OrAssign => Some(AssignmentOperator::LogicalOr),
            Token::BitAndAssign => Some(AssignmentOperator::BitAnd),
            Token::AndAssign => Some(AssignmentOperator::LogicalAnd),
            Token::XorAssign => Some(AssignmentOperator::BitXor),
            Token::ShiftLAssign => Some(AssignmentOperator::BitShiftLeft),
            Token::ShiftRAssign => Some(AssignmentOperator::BitShiftRight),
            Token::NullCoalesceAssign => Some(AssignmentOperator::NullCoalesce),
            _ => None,
        };
        if let Some(assignment_operator) = assignment {
            self.bump();
            let rhs = self.expression_node()?;
            Ok(self.node_at(
                lhs.start,
                rhs.end,
                AssignmentExpression {
                    left: lhs,
                    operator: assignment_operator,
                    right: rhs,
                },
            ))
        } else {
            Ok(lhs)
        }
    }

    // Right-associative
    fn conditional_expression(&mut self) -> Result<ExpressionNode<'ast>> {
        let condition = self.comparing_expression()?;
        if self.eat(Token::Question) {
            let true_value = self.expression_node()?;
            self.expect(Token::Colon);
            let false_value = self.expression_node()?;
            Ok(self.node_at(
                condition.start,
                false_value.end,
                TernaryExpression {
                    condition,
                    when_true: true_value,
                    when_false: false_value,
                },
            ))
        } else {
            Ok(condition)
        }
    }

    // Right-associative
    fn comparing_expression(&mut self) -> Result<ExpressionNode<'ast>> {
        let lhs = self.comparand()?;
        let operator = match self.current_token {
            Token::Equals => Some(BinaryOperator::Equals),
            Token::GreaterThan => Some(BinaryOperator::GreaterThan),
            Token::GreaterThanEquals => Some(BinaryOperator::GreaterThanEquals),
            Token::NotEquals => Some(BinaryOperator::NotEquals),
            Token::LessThan => Some(BinaryOperator::LessThan),
            Token::LessThanEquals => Some(BinaryOperator::LessThanEquals),
            Token::In => Some(BinaryOperator::In),
            _ => None,
        };
        if let Some(binary_operator) = operator {
            self.bump();
            let rhs = self.expression_node()?;
            Ok(self.node_at(
                lhs.start,
                rhs.end,
                BinaryExpression {
                    left: lhs,
                    operator: binary_operator,
                    right: rhs,
                },
            ))
        } else {
            Ok(lhs)
        }
    }

    // Left-associative
    fn comparand(&mut self) -> Result<ExpressionNode<'ast>> {
        let lhs = self.term()?;
        let mut expr = lhs;
        loop {
            let operator = match self.current_token {
                Token::Add => Some(BinaryOperator::Addition),
                Token::Sub => Some(BinaryOperator::Subtraction),
                Token::Or => Some(BinaryOperator::LogicalOr),
                Token::BitOr => Some(BinaryOperator::BitOr),
                _ => None,
            };
            if let Some(binary_operator) = operator {
                self.bump();
                let rhs = self.term()?;
                expr = self.node_at(
                    expr.start,
                    expr.end,
                    BinaryExpression {
                        left: lhs,
                        operator: binary_operator,
                        right: rhs,
                    },
                );
            } else {
                break Ok(expr);
            }
        }
    }

    // Left-associative
    fn term(&mut self) -> Result<ExpressionNode<'ast>> {
        let lhs = self.factor()?;
        let mut expr = lhs;
        loop {
            let operator = match self.current_token {
                Token::Mul => Some(BinaryOperator::Multiplication),
                Token::Quo => Some(BinaryOperator::Division),
                Token::And => Some(BinaryOperator::LogicalAnd),
                Token::BitAnd => Some(BinaryOperator::BitAnd),
                _ => None,
            };
            if let Some(binary_operator) = operator {
                self.bump();
                let rhs = self.factor()?;
                expr = self.node_at(
                    expr.start,
                    expr.end,
                    BinaryExpression {
                        left: lhs,
                        operator: binary_operator,
                        right: rhs,
                    },
                );
            } else {
                break Ok(expr);
            }
        }
    }

    // Left-associative
    fn factor(&mut self) -> Result<ExpressionNode<'ast>> {
        let lhs = self.unary()?;
        let mut expr = lhs;
        loop {
            let operator = match self.current_token {
                Token::ShiftL => Some(BinaryOperator::BitShiftLeft),
                Token::ShiftR => Some(BinaryOperator::BitShiftRight),
                _ => None,
            };
            if let Some(binary_operator) = operator {
                self.bump();
                let rhs = self.unary()?;
                expr = self.node_at(
                    expr.start,
                    expr.end,
                    BinaryExpression {
                        left: lhs,
                        operator: binary_operator,
                        right: rhs,
                    },
                );
            } else {
                break Ok(expr);
            }
        }
    }

    fn unary(&mut self) -> Result<ExpressionNode<'ast>> {
        let operator = match self.current_token {
            Token::Not => Some(PrefixOperator::LogicalNot),
            Token::BitNot => Some(PrefixOperator::BitNot),
            Token::Sub => Some(PrefixOperator::Minus),
            Token::Add => Some(PrefixOperator::Plus),
            Token::Increment => Some(PrefixOperator::Increment),
            Token::Decrement => Some(PrefixOperator::Decrement),
            _ => None,
        };
        if let Some(prefix_operator) = operator {
            let start = self.loc().0;
            self.bump();
            let rhs = self.unary()?;
            return Ok(self.node_at(
                start,
                rhs.end,
                PrefixExpression {
                    operator: prefix_operator,
                    operand: rhs,
                },
            ));
        }

        let mut expr = self.primary_expression()?;
        loop {
            let operator = match self.current_token {
                Token::Increment => Some(PostfixOperator::Increment),
                Token::Decrement => Some(PostfixOperator::Decrement),
                Token::Not => Some(PostfixOperator::NullForgiving),
                _ => None,
            };
            if let Some(postfix_operator) = operator {
                let end = self.loc().1;
                self.bump();
                expr = self.node_at(
                    expr.start,
                    end,
                    PostfixExpression {
                        operator: postfix_operator,
                        operand: expr,
                    },
                );
            } else {
                break Ok(expr);
            }
        }
    }

    fn primary_expression(&mut self) -> Result<ExpressionNode<'ast>> {
        match self.current_token {
            Token::Integer(_) => Ok(self.node_from_slice(|s| Primitive::DecimalNumber(s))),
            Token::LiteralString => Ok(self.node_from_slice(|s| Primitive::String(s))),
            Token::Bool(b) => Ok(self.node_at_token(Primitive::Bool(b))),
            Token::Null => Ok(self.node_at_token(Primitive::Null)),
            Token::LParen => {
                self.expect(Token::LParen);
                if self.peek_token == Token::Colon {
                    // TODO: Lambdas
                    Err(Error::NotImplementedError)
                } else {
                    let inside = self.expression_node()?;
                    self.expect(Token::RParen);
                    Ok(inside)
                }
            }
            _ => self.reference(),
        }
    }

    fn reference(&mut self) -> Result<ExpressionNode<'ast>> {
        // TODO: Remove this exception
        #[allow(clippy::match_single_binding)]
        match self.current_token {
            // Token::At => Address-of reference
            // Token::Mul => Dereference
            // TODO: Member access
            // TODO: Function call
            // TODO: Constructor call
            // TODO: Array reference
            // TODO: Cast reference
            _ => {
                let ident = self.identifier_node()?;
                Ok(self.node_at(ident.start, ident.end, ident))
            }
        }
    }
}
