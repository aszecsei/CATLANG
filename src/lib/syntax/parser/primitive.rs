use crate::syntax::ast::*;
use crate::syntax::error::*;
use crate::syntax::lexer::Token;
use crate::syntax::parser::Parser;

impl<'ast> Parser<'ast> {
    pub fn identifier_node(&mut self) -> Result<IdentifierNode<'ast>> {
        let (start, end) = (self.current_span.start as u32, self.current_span.end as u32);
        let val = self.current_slice;
        self.expect(Token::Ident);
        Ok(self.node_at(start, end, val))
    }

    pub fn string_literal_node(&mut self) -> Result<StringLiteralNode<'ast>> {
        let (start, end) = (self.current_span.start as u32, self.current_span.end as u32);
        let val = self.current_slice;
        self.expect(Token::LiteralString);
        Ok(self.node_at(start, end, val))
    }
}
