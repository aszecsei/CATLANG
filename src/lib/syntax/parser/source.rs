use toolshed::list::GrowableList;

use crate::syntax::ast::*;
use crate::syntax::error::*;
use crate::syntax::lexer::Token;
use crate::syntax::parser::Parser;

impl<'ast> Parser<'ast> {
    pub fn source_unit(&mut self) -> SourceUnitNode<'ast> {
        let start = self.current_span.start as u32;
        let module = Module {
            elements: GrowableList::new().as_list(),
        };
        let end = self.last_span.end as u32;
        self.node_at(start, end, module)
    }
}

fn is_declaration_starter(t: Token) -> bool {
    match t {
        Token::Export => true,
        Token::Const => true,
        Token::Type => true,
        Token::Let => true,
        Token::Function => true,
        Token::Struct => true,
        Token::Enum => true,
        _ => false,
    }
}
