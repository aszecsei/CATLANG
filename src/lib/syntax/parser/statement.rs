use toolshed::list::ListBuilder;

use crate::syntax::ast::*;
use crate::syntax::error::*;
use crate::syntax::lexer::Token;
use crate::syntax::parser::Parser;

impl<'ast> Parser<'ast> {
    pub fn statement_node(&mut self) -> Result<StatementNode<'ast>> {
        Err(Error::NotImplementedError)
    }
}
