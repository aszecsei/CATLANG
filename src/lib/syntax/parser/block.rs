use toolshed::list::GrowableList;

use crate::syntax::ast::*;
use crate::syntax::error::*;
use crate::syntax::lexer::Token;
use crate::syntax::parser::module::is_declaration_starter;
use crate::syntax::parser::Parser;

impl<'ast> Parser<'ast> {
    pub fn block_node(&mut self) -> Result<BlockNode<'ast>> {
        let start = self.loc().0;
        self.expect(Token::LCurlyB);
        let elements = GrowableList::new();
        loop {
            match self.current_token {
                Token::RCurlyB => break,
                Token::Semicolon => self.bump(),
                _ => elements.push(self.arena, self.block_element()?),
            }
        }
        let end = self.loc().1;
        self.expect(Token::RCurlyB);
        Ok(self.node_at(
            start,
            end,
            Block {
                elements: elements.as_list(),
            },
        ))
    }

    fn block_element(&mut self) -> Result<BlockElementNode<'ast>> {
        if self.current_token == Token::Import {
            let import = self.import()?;
            Ok(self.node_at(import.start, import.end, import))
        } else if is_declaration_starter(self.current_token) {
            let declaration = self.declaration_node()?;
            Ok(self.node_at(declaration.start, declaration.end, declaration))
        } else {
            let statement = self.statement_node()?;
            Ok(self.node_at(statement.start, statement.end, statement))
        }
    }
}
