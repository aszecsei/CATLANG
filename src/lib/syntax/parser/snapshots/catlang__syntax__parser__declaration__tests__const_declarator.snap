---
source: src/lib/syntax/parser/declaration.rs
expression: res
---
(0:12) Declaration {
    attributes: [],
    declarator: Constant(
        ConstantDeclarator {
            identifier: (6:8) "PI",
            type_expression: None,
            expression: (11:12) PrimitiveExpression(
                DecimalNumber(
                    "3",
                ),
            ),
        },
    ),
}
